### PR TITLE
Add the option to specify custom directories with user-defined profiles

### PR DESCRIPTION
**Pull request by zacikpa**
The new default directory for user-defined profiles is `/etc/tuned/profiles/`; other directories can be specified via the `custom_profile_dirs` option in `tuned-main.conf`.

When updating the TuneD package, existing profiles in `/etc/tuned/` are automatically copied to `/etc/tuned/profiles/`, unless there are conflicts.

GUI also needed to be updated - I only did the bare minimum for it to work, more major refactoring deserves a separate PR.

### CODE REVIEW

File: tuned/consts.py
Comment (jmencak) [inline tuned/consts.py:125]: As mentioned above, it might be a good idea to change `_dirs` to `_paths` due to consistency with other `tuned-main.conf` options.

**Review by jmencak — COMMENTED** [review]
Thank you for this PR, Pavol, I very much appreciate it!  I tested the following worked::
* profile migration script
* specifying custom directories in `tuned-main.conf` -- profiles applied from custom dirs

I did not test the GUI though.

The code looks clean to me.  I found a couple of nits and the only real suggestion for improvement I have to add the new option commented out in `tuned-main.conf`

File: tuned.spec
Comment (zacikpa) [reply tuned.spec]: Agreed,  changed.

File: doc/manual/modules/performance/con_the-location-of-tuned-profiles.adoc
Comment (zacikpa) [reply doc/manual/modules/performance/con_the-location-of-tuned-profiles.adoc:18]: Thanks, fixed.

File: doc/manual/modules/performance/con_the-location-of-tuned-profiles.adoc
Comment (zacikpa) [reply doc/manual/modules/performance/con_the-location-of-tuned-profiles.adoc]: Done.

File: tuned.spec
Comment (zacikpa) [reply tuned.spec]: I agree with all three points, updated to use just `-e`.

File: tuned/consts.py
Comment (zacikpa) [reply tuned/consts.py:125]: Here I would disagree. The documentation/manual pages always refer to (custom) profile directories, not paths, so directories seem more clear to me.

**Comment by zacikpa** [comment]
Thanks, @jmencak, for the thorough review. I now updated the PR.

**Comment by jmencak** [comment]
Hi @zacikpa , thank you for the effort, I really appreciate it.  I've just realized a blocker for this from OpenShift side.  While the migration script will work on RHEL and on upgrades, it will not work in container environments where the profiles are still expected in /etc/tuned and the supervising process is unpacking them to /etc/tuned.  An FDP release of this would immediately break all OCP versions which used FDP (OCP 4.10+).  There is unfortunately no way to adjust the supervising process which extracts TuneD profiles all the way down to 4.10.  I do not have a good solution for this at the moment apart from still supporting the obsoleted /etc/tuned for the profiles.  One idea I have could be to ship with the following `tuned-main.conf` file:

```
custom_profile_dirs=/etc/tuned/profiles,/etc/tuned
```

While I do understand this is not clean from the TuneD/RHEL perspective, I currently do not have a better solution to this for the OCP environment.  Thoughts? cc @yarda

**Comment by zacikpa** [comment]
A more extreme solution would be to keep `/etc/tuned/` as the default directory but still merge the changes that would allow users to configure it.

**Comment by jmencak** [comment]
> A more extreme solution would be to keep `/etc/tuned/` as the default directory but still merge the changes that would allow users to configure it.

Yes, this is also a possibility with some INFO (or even WARNING) message that the support for profile directory '/etc/tuned' will be dropped in future versions. Not sure which approach is better to be honest.  I'll let you/Jarda to decide.

**Comment by yarda** [comment]
There will be probably no further RHEL-8 TuneD release apart from the FDP, thus we cannot change the default directory in RHEL-8. I am also afraid it could break things.

For RHEL-9, I am also afraid it could break things for some people. 

I.e. I would change the profile location in upstream, i.e. in this PR and in further RHELs (e.g. RHEL-10). For older RHELs I would just backport the option from upstream, but wouldn't change the defaults. @jmencak if you want to change it in OCP you could reconfigure TuneD.

File: tuned.spec
Comment (yarda) [inline tuned.spec:334]: IMHO the current system profile directory is `/usr/lib/tuned/`, not `/usr/lib/tuned/profiles/`.

But I think you should migrate all profiles, because there could be custom profiles overrides in `/etc/tuned/` with the same name as profiles in `/var/lib/tuned/`. It's quite common to e.g. do:
```
# custom throughput-performance profile in /etc/tuned overriding setting of throughput-performance profile in /var/lib/tuned
[main]
# the stock one
include=throughput-performance

# customization
...
```

**Comment by jmencak** [comment]
> There will be probably no further RHEL-8 TuneD release apart from the FDP, thus we cannot change the default directory in RHEL-8. I am also afraid it could break things.
> 
> For RHEL-9, I am also afraid it could break things for some people.
> 
> I.e. I would change the profile location in upstream, i.e. in this PR and in further RHELs (e.g. RHEL-10). For older RHELs I would just backport the option from upstream, but wouldn't change the defaults. @jmencak if you want to change it in OCP you could reconfigure TuneD.

Reconfiguring TuneD is a problem for very old but still supported OCP releases I'm afraid. As soon as you merge this and create and FDP out of it (as is), things will break in OCP.

**Comment by yarda** [comment]
> > There will be probably no further RHEL-8 TuneD release apart from the FDP, thus we cannot change the default directory in RHEL-8. I am also afraid it could break things.
> > For RHEL-9, I am also afraid it could break things for some people.
> > I.e. I would change the profile location in upstream, i.e. in this PR and in further RHELs (e.g. RHEL-10). For older RHELs I would just backport the option from upstream, but wouldn't change the defaults. @jmencak if you want to change it in OCP you could reconfigure TuneD.
> 
> Reconfiguring TuneD is a problem for very old but still supported OCP releases I'm afraid. As soon as you merge this and create and FDP out of it (as is), things will break in OCP.

We will just add the option and don't change the defaults in older RHELs.  With the proposed:
```
profile_dirs = /usr/lib/tuned/profiles;/etc/tuned/profiles
```
You could reconfigure it yourself in OCP as you would like, e.g.:
```
profile_dirs = /usr/lib/tuned;/usr/lib/tuned/profiles;/etc/tuned;/etc/tuned/profiles
```
With such config all dirs will be supported and the upstream default dirs will have higher priority to override.

**Comment by yarda** [comment]
Also the new upstream dirs will be seen only in OCP, not in standard RHEL-8/9, which will have the same config as it have now.

File: tuned.spec
Comment (zacikpa) [reply tuned.spec:334]: The second check is actually only there to make sure that we don't overwrite existing directories in `/etc/tuned/profiles/` - though they have no reason to exist before the migration, and I could also do that more cleanly using `mv -n`.

**Comment by zacikpa** [comment]
> @zacikpa I think the system profiles should be also moved to `/usr/lib/tuned/profiles`. Because, currently there can be the same problem with the conflicting profile names and special files/configuration files in this directory.

Ok, I will do the migration of both custom and system profiles in a separate PR. 

> Maybe global configuration option for it would be also good, but I think we don't need to auto-migrate them (the system profiles).
> 
> I am not sure whether two configuration options (custom/system profiles) have any benefit, maybe one configuration option for all? E.g.:
> 
> ```
> profile_dirs = /usr/lib/tuned/profiles;/etc/tuned/profiles
> ```
> 
> With the priority from left to right? The rightmost, the most prio. Opinions?

I think having it in a single option would be quite clean. Tuned-gui needs to explicitly distinguish the system profiles from the custom ones, however, so I'll have to think about how to resolve that.

**Comment by zacikpa** [comment]
I'm also in favor of specifying the default values via `tuned.consts` and not directly via `tuned-main.conf` because the migration might then break TuneD for users that already use a modified version of `tuned-main.conf.`

**Comment by yarda** [comment]
> I'm also in favor of specifying the default values via `tuned.consts` and not directly via `tuned-main.conf` because the migration might then break TuneD for users that already use a modified version of `tuned-main.conf.`

I think it may work as with other options. If unspecified or commented out then the default values from the consts should be used. For the auto migration you could check whether it's supported configuration and if not then do not auto-migrate. I think profiles under `/usr/lib/tuned` shouldn't be auto-migrated, just profiles in the '/etc/tuned' should be auto-migrated. I think profiles under `/usr/lib/tuned` should be handled by distro maintainer and if some users put their profiles there it's their responsibility to cope with it.

**Comment by zacikpa** [comment]
The PR now contains only changes related to the new global configuration option, `profile_dirs`. Please re-review. I will open a new PR for the migration to new default profile locations.

File: tuned-main.conf
Comment (yarda) [inline tuned-main.conf:88]: I would write "has priority" or "takes precedence", but I am not native speaker.

File: tuned/gtk/gui_profile_loader.py
Comment (yarda) [inline tuned/gtk/gui_profile_loader.py:63]: I think this could be configurable, maybe a list with system profile dirs? In tuned-gtk.conf? Or tuned-main.conf? It could be subset of profile_dirs that would be treated as as system profile dirs.

Or maybe a better way: Add tuned-gtk.conf and option like profile_save_dir which specifies where to save user profiles.

But this can be handled in later PRs, so LGTM for this PR.

**Review by yarda — APPROVED** [review]
Thanks, LGTM. Maybe improve the comment as mentioned earlier.

File: tuned-main.conf
Comment (zacikpa) [reply tuned-main.conf]: Thanks, fixed. I mixed up the two.

File: tuned/gtk/gui_profile_loader.py
Comment (zacikpa) [reply tuned/gtk/gui_profile_loader.py:63]: I agree. In fact, I have many more refactoring changes prepared for the GUI. To make this PR easily reviewable, I only included the necessary changes that keep the GUI working (more or less) as before.

**Comment by yarda** [comment]
LGTM,thanks.

**Pull request merged**
